api/v1alpha1/gateway_types.go

```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
 type GatewaySpec struct {
 	// Class used for this Gateway. This is the name of a GatewayClass resource.
 	Class string `json:"class" protobuf:"bytes,1,opt,name=class"`
-	// Listeners associated with this Gateway. Listeners define what addresses,
-	// ports, protocols are bound on this Gateway.
+
+	// Listeners associated with this Gateway. Listeners define logical
+	// endpoints that are bound on this Gateway. If no Listeners are
+	// provided, the GatewayClass may associate the Gateway with
+	// an address but the Gateway will not be able to accept any
+	// connections.
+	//
+	// +optional
 	Listeners []Listener `json:"listeners" protobuf:"bytes,2,rep,name=listeners"`
-	// Routes specifies a schema for associating routes with the Gateway using
-	// selectors. A route is a resource capable of servicing a request and allows
-	// a cluster operator to expose a cluster resource (i.e. Service) by
-	// externally-reachable URL, load-balance traffic and terminate SSL/TLS.
-	// Typically, a route is a "httproute" or "tcproute" in group
-	// "networking.x-k8s.io". However, an implementation may support other resources.
+
+	// Address requested for this gateway. This is optional and behavior
+	// can depend on the GatewayClass. If a value is set in the spec and
+	// the requested address is invalid, the GatewayClass MUST indicate
+	// this in the associated entry in GatewayStatus.Listeners. If no
+	// addresses are specified, the GatewayClass may schedule the Gateway
+	// in an implementation-defined manner, providing it with an appropriate
+	// set of addresses.
 	//
-	// Support: Core
+	// Support:
 	//
-	Routes RouteBindingSelector `json:"routes" protobuf:"bytes,3,opt,name=routes"`
+	// +optional
+	Addresses []ListenerAddress `json:"endpoints" protobuf:"bytes,3,rep,name=endpoints"`
 }
 
+// ProtocolType defines the application protocol accepted by a Listener.
+type ProtocolType string
+
+const (
+	// HTTPProtocolType accepts cleartext HTTP/1.1 sessions over TCP.
+	HTTPProtocolType ProtocolType = "HTTP"
+
+	// HTTPSProtocolType accepts HTTP/1.1 or HTTP/2 sessions over TLS.
+	HTTPSProtocolType ProtocolType = "HTTPS"
+
+	// TLSProtocolType accepts TLS sessions over TCP.
+	TLSProtocolType ProtocolType = "TLS"
+
+	// TCPProtocolType accepts TCP sessions.
+	TCPProtocolType ProtocolType = "TCP"
+)
+
+// HostnameMatchType specifies the types of matches that are valid
+// for host names.
+type HostnameMatchType string
+
 const (
-	// HTTPProtocol constant.
-	HTTPProtocol = "HTTP"
-	// HTTPSProtocol constant.
-	HTTPSProtocol = "HTTPS"
+	// HostnameMatchDomain specifies that the host name provided by the
+	// client should be matched against a DNS domain value. The domain
+	// match should follow the wildcard certificate matching rules in
+	// [RFC 6125](https://tools.ietf.org/html/rfc6125#section-6.4.3).
+	HostnameMatchDomain HostnameMatchType = "Domain"
```

Comment:
As a user, `Domain` seems to be as confusing as `Local` and `Cluster` external traffic policies, `domain` is so generic that it can mean different things to different people.
How about `Wildcard`?